generate-fit-profile/src/decode.rs

```diff
@@ -50,19 +50,33 @@ fn decode_function_def(msg: &MessageDefinition) -> TokenStream {
 
     quote! {
         #(#comments)*
-        fn #fn_name(mesg_num: MesgNum, data_map: &mut HashMap<u8, Value>, accumlators: &mut HashMap<u32, Value>, options: &HashSet<DecodeOption>) -> Result<Vec<FitDataField>> {
+        fn #fn_name(mesg_num: MesgNum, data_map: &mut HashMap<u8, Value> , developer_data_map: &HashMap<(u8, u8), Value>, accumlators: &mut HashMap<u32, Value>, developer_field_descriptions: &HashMap<(u8, u8), DeveloperFieldDescription>, options: &HashSet<DecodeOption>) -> Result<Vec<FitDataField>> {
             let mut fields = Vec::new();
             let mut entries: VecDeque<(u8, Value)> = data_map.iter().map(|(k, v)| (*k, v.clone())).collect();
-            while let Some((def_num, value)) = entries.pop_front() {
-                match def_num {
+            while let Some((field_nr, value)) = entries.pop_front() {
+                match field_nr {
                     #(#match_arms)*
                     _ => {
                         if !options.contains(&DecodeOption::DropUnknownFields) {
-                            fields.push(unknown_field(def_num, value));
+                            fields.push(unknown_field(field_nr, value));
                         }
                     }
                 }
             }
+            let mut entries: VecDeque<((u8, u8), Value)> = developer_data_map.iter().map(|(k, v)| (*k, v.clone())).collect();
+            while let Some(((dev_data_idx, field_nr), value)) = entries.pop_front() {
+                let dev_definition = developer_field_descriptions.get(&(dev_data_idx, field_nr)).ok_or(ErrorKind::MissingDeveloperDefinitionMessage())?;
+                fields.push(data_field_with_info(
+                    dev_definition.field_definition_number,
+                    &dev_definition.field_name,
```

Comment:
ISTM that developer fields are identified only by their developer-provided name in the resultant `FitDataField` and by the field definition number. This makes it hard to identify which application a field comes from and makes it inconvenient or hard to deal with developers that provide a field with a name that matches a generic field name.

Would it make sense to store part of FitDataField in an Either/enum? The ways to identify a base field (name or number that's unique within all messages of a given MesgNum) and a developer field (app id *together with* a name, the number not being meaningful and changing across different fit files with the field present) are totally different, after all.

Reply:
@robryk I wonder if the intent for a developer field with the same name is to "override" that field? It is a bit of a hangup but not sure if it would be worth an additional layer of wrapping on access?

Reply:
There's a different way to comminucate that intent (`native_field_num`, which IIUC might be absent). I'd argue that they shouldn't override that field either: doing so makes it impossible to actually access the overridden thing (e.g. if you want to compare the two sources of supposedly-matching data) or to know that it was overridden.

Reply:
I am in favor of the first approach, because the way you identify a dev field is different. To uniquely identify a "common" field, you need its canonical name. For a dev field, a unique identification consists of its name _and the app UUID_.

Reply:
@robryk I agree, give it a look now, we should be good to go on separating the data. @Lingepumpe do you also mind testing this? I think once we're in agreement here this will be ready to merge.

Reply:
Sorry, I was unavailable the last weeks, I am back now and will take a look at this tomorrow.

Reply:
I have tested the changes, and can confirm that on my fit files the developer fields are correctly extracted (only now not in fields, but in developer_fields).

I don't think splitting into fields and developer_fields is a positive decision from the perspective of a user of the library:
- Distinguishing developer vs non developer fields is a very technical/specific detail that somebody who wants to "parse the fit file" should not be forced to learn about.
- Other libraries agree, combining the developer and normal fields in the output they return to the library user, e.g. https://github.com/dtcooper/python-fitparse 
- The concern raised that there could be duplication with the exact same field name being reused is valid according to the fitfile specifications, but all developer fields I know have intentionally avoided reusing the a native field name (e.g. Stryd foot-pod uses "Power" vs "power" and "Cadence" vs "cadence" etc.). My assumption is that the recording head unit (e.g. watch) might refuse a developer field that tries to re-use a existing name, or that at the very minimum 3rd party applications would break [e.g. anything using the above linked python-fitparse]. So I think the concern is only hypothetical.
- The solution of splitting into fields and developer_fields requires extra code for the normal "default" case of trying to extract all the fields of fit file. E.g., even the example `fit_to_json` utility included in `fitparse-rs` currently doesn't show the developer_fields with this change (it did before splitting).

Reply:
If we want to be sure that a user of the library can identify the source of a field, then we should not only show if it is a developer_field or a "built in" field, as the current status of this branch does, but rather provide the information about *which* application_id (a  16-byte identifier for the developer) or developer_data_index was responsible for each field. For built-in fields this could be None, e.g.

```rust
pub struct FitDataRecord {
    /// The kind of message the data came from, the FIT profile defines several messages and
    /// custom messages can be defined by altering the profile
    kind: profile::MesgNum,
    /// All the fields present in this message, a record may not have every possible field defined
    fields: Vec<FitDataField>,
}

pub struct FitDataField {
    name: String,
    number: u8,
    developer_data_index: Option<u8>, // None for built-in fields, otherwise the developer_data_index
    value: Value,
    units: String,
}
```

Reply:
I have pushed a version with this suggestion to the branch. This allows library users to track exactly which field is built-in, which is a developer field and if so, from which developer did it come from.
It also does not force the knowledge of the difference between developer fields and normal fields on the library user, since the fields now contain all the fields (no matter the source), for those that don't care about this detail. This also means the example `fit_to_json` now includes all the developer fields again, which I believe is the desired behavior.

Reply:
> but rather provide the information about which application_id (a 16-byte identifier for the developer) or developer_data_index was responsible for each field.

IIUC the developer_data_index is not useful for identification (I was under the impression that it's just sequentially assigned to all developer fields present in a fit file) and the way to identify a developer field uniquely is by the (application_id, name) pair.

Reply:
I think your reasoning for not splitting the fields is good. Personally, I think I’d rather have them split (hence my change), but from a library API perspective it’s reasonable and simpler to just make them identifiable. The ability to trace back to with particular developer encoded it is also a good feature. 

I do wonder if we’d want to replace the developer data index with the actual application_id. We don’t report the local message numbers and only return the global message numbers since that is file specific data. In the same sense here, the dev index is file local data and not particularly useful to the end user. Reporting the actual app_id means they can do direct value comparison instead of collecting the index from a specific message and comparing index values.

Reply:
The reason why I used the `developer_data_index` instead of the `application_id` in each field, is the size: application_id is 16 bytes and developer_data_index is only 1 byte. As there can easily be 10k records in a fit file, and having 10 developer fields per record is happening in my example, that would be 1.6 MB of extra RAM, which is not huge but also not nothing.

Speculating on what `fitparse-rs` users might want:
- Most users will be fine just using the name to identify a field. I have not seen a single real world fit file where there was indeed name overlap, and as pointed out, other well used fitparse libraries also only give you the name, and there hasn't been an outcry of people needing more info.
- Of the few users that are not fine with just using the name, that want to correctly parse even a fitfile that has name overlap (if it exists!), most of these will just want to be able to group the data of the fitfile correctly (e.g. if you want to create a spreadsheet with the fit file data, you need to identify the column the data belongs to; Or if you want to create plots for all the data in a fit file [similar to what Garmin Connect does when you open a fitfile], then you need to group the data to create each plot only with data that belongs together). All of this can be done with the `(name, developer_data_index)` as identifier, and does not require the application ID.
- This leaves us with very few, if any, users that will want to know which exact application_id a developer field comes from. Since I assume this to be such a niche scenario, I thought it fine to require a "two step process" for this use case, where you would store the `(developer_data_index, application_id)` mapping yourself.

Options (in order of my preference):
- Keep it as is, requiring users that care about application_id to remember the mapping and look it up when they encounter a field.
- Provide a function to look-up for the application-id based on developer_data_index. This would just mean we keep track of the mapping inside `fitfile-rs`, to make it easier (but still a two-step process) to resolve a `developer_data_index` to a `application_id`.
- Add the application_id to every field (and accept the size increase coming with that)

Reply:
I think that reasoning sounds good to me to leave it as-is until there is enough of a push to do something more. I hadn't really considered the size increase for large FIT files and while I can't imagine that anyone is using this on an embedded device or just at absurd scales where that 1.6Mb would actually matter we don't gain/lose much either way in terms of added functionality with better handling of the app ids.